### ISSUE TITLE
AddDefaultEndpointSelectorPolicy did not work with the new MicrosoftDI.GraphQLBuilderExtensions

### ISSUE THREAD

**Issue by lanwin** [graphql-dotnet/server#705]
Hi, I am currently upgrading our solution to the latest Graphql Version and found a problem.

Ive upgraded the solution, started it and it works. But I got the AddGraphQl obsolete warnings (by the way, it would be helpful if they had a description what todo to get rid of the obsolete).

So ive changed: 

```csharp
  services.AddGraphQL((options, provider) =>
  {
      options.EnableMetrics = true;
      var logger = provider.GetRequiredService<ILogger<Startup>>();
      options.UnhandledExceptionDelegate = ctx => logger.LogError("{Error} occurred", ctx.OriginalException.Message);
  })
  .AddDefaultEndpointSelectorPolicy()
  .AddSystemTextJson(deserializerSettings => { }, serializerSettings => { })
  .AddWebSockets()
  .AddDataLoader()
  .AddGraphTypes(typeof(MySchema));
```

to 

```csharp
    GraphQL.MicrosoftDI.GraphQLBuilderExtensions.AddGraphQL(services)
    .AddServer(true,
        (options, provider) =>
        {
            options.EnableMetrics = true;
            var logger = provider.GetRequiredService<ILogger<Startup>>();
            options.UnhandledExceptionDelegate = ctx => logger.LogError("{Error} occurred", ctx.OriginalException.Message);
        })
    .AddDefaultEndpointSelectorPolicy()
    .AddSystemTextJson(deserializerSettings => { }, serializerSettings => { })
    .AddWebSockets()
    .AddGraphTypes(typeof(MySchema).Assembly);
```

And now Ive get:
```
Microsoft.AspNetCore.Routing.Matching.AmbiguousMatchException: The request matched multiple endpoints. Matches:
GraphQL WebSockets
GraphQL
```

Using NET5 and Routing.

Ive played around with the various extensions, but did not find a working solution yet.

**Comment by sungam3r** [comment]
Did you add `AddRouting`?

**Comment by lanwin** [comment]
> `AddRouting`

Yes three lines above AddGraphQL.             

```
services.AddRouting();
services.AddResponseCompression();
```

**Comment by sungam3r** [comment]
To investigate further I need to see `ConfigureServices` and `Configure` methods.

**Comment by sungam3r** [comment]
Or you may copy source code, rename class and just debug it:
```c#
 internal sealed class MyGraphQLDefaultEndpointSelectorPolicy : MatcherPolicy, IEndpointSelectorPolicy
    {
        public override int Order => int.MaxValue;

        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
        {
            for (int i = 0; i < endpoints.Count; ++i)
            {
                if (endpoints[i].DisplayName == "GraphQL" || endpoints[i].DisplayName == "GraphQL WebSockets")
                    return true;
            }

            return false;
        }

        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
        {
            if (candidates.Count < 2)
                return Task.CompletedTask;

            for (int i = 0; i < candidates.Count; ++i)
            {
                if (!candidates.IsValidCandidate(i))
                    continue;

                ref var state = ref candidates[i];

                if (state.Endpoint.DisplayName == "GraphQL" && httpContext.WebSockets.IsWebSocketRequest)
                {
                    candidates.SetValidity(i, false);
                }

                if (state.Endpoint.DisplayName == "GraphQL WebSockets" && !httpContext.WebSockets.IsWebSocketRequest)
                {
                    candidates.SetValidity(i, false);
                }
            }

            return Task.CompletedTask;
        }
    }
```

**Comment by sungam3r** [comment]
Register `MyGraphQLDefaultEndpointSelectorPolicy` as singleton.

**Comment by sungam3r** [comment]
`AddDefaultEndpointSelectorPolicy` is a thin wrapper around DI registration:

```c#
public static DI.IGraphQLBuilder AddDefaultEndpointSelectorPolicy(this DI.IGraphQLBuilder builder)
        {
            builder.TryRegister<MatcherPolicy, GraphQLDefaultEndpointSelectorPolicy>(DI.ServiceLifetime.Singleton);

            return builder;
        }
```

**Comment by sungam3r** [comment]
Ooops, I reproduced this bug.

**Comment by lanwin** [comment]
Interesting. If I use `services.AddSingleton<MatcherPolicy, MyGraphQLDefaultEndpointSelectorPolicy>(); ` right after, it starts to work again.

**Comment by lanwin** [comment]
Not sure why. The decompiled version of GraphQLDefaultEndpointSelectorPolicylooks exactly like the version you send me above.

**Comment by lanwin** [comment]
Can it be TryRegister? If I understand it correctly it checks to MatcherPolicy and not GraphQLDefaultEndpointSelectorPolicy. So Ive there is already a MatcherPolicy registered. The GraphQLDefaultEndpointSelectorPolicy did not get registered.

**Comment by sungam3r** [comment]
@Shane32 It's a bug. 
Before `IGraphQLBuilder` rework I see this line:
```
 builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<MatcherPolicy, GraphQLDefaultEndpointSelectorPolicy>());
```

**Comment by sungam3r** [comment]
Enumerable registration is a special case.

**Comment by sungam3r** [comment]
So... We need to review all registrations again to find all _single_ registrations and _enumerable_ registrations. Perhaps, public API should be changed also.

**Comment by sungam3r** [comment]
TryAddEnumerable:
```c#
 public static void TryAddEnumerable(
            this IServiceCollection services,
            ServiceDescriptor descriptor)
        {
            if (services == null)
            {
                throw new ArgumentNullException(nameof(services));
            }

            if (descriptor == null)
            {
                throw new ArgumentNullException(nameof(descriptor));
            }

            Type? implementationType = descriptor.GetImplementationType();

            if (implementationType == typeof(object) ||
                implementationType == descriptor.ServiceType)
            {
                throw new ArgumentException(
                    SR.Format(SR.TryAddIndistinguishableTypeToEnumerable,
                        implementationType,
                        descriptor.ServiceType),
                    nameof(descriptor));
            }

            if (!services.Any(d =>
                              d.ServiceType == descriptor.ServiceType &&
                              d.GetImplementationType() == implementationType))
            {
                services.Add(descriptor);
            }
        }
```
vs TryAdd:
```c#
 public static void TryAdd(
            this IServiceCollection collection,
            ServiceDescriptor descriptor)
        {
            if (collection == null)
            {
                throw new ArgumentNullException(nameof(collection));
            }

            if (descriptor == null)
            {
                throw new ArgumentNullException(nameof(descriptor));
            }

            if (!collection.Any(d => d.ServiceType == descriptor.ServiceType))
            {
                collection.Add(descriptor);
            }
        }
```

**Comment by Shane32** [comment]
Makes sense. So we just need to add TryRegisterEnumerable, right? I can make a PR

**Comment by sungam3r** [comment]
Only `AddDefaultEndpointSelectorPolicy` uses TryAddEnumerable.

**Comment by sungam3r** [comment]
> So we just need to add TryRegisterEnumerable, right?

I would not. Wait a minute, I'll try to provide a public API changes.

**Comment by Shane32** [comment]
We can just cast to the underlying ServiceCollection and use TryAddEnumerable from there.

**Comment by Shane32** [comment]
Server is already tied to MS DI for authentication

**Comment by Shane32** [comment]
Also since you have to manually call the builder method it’s not like it may have been called twice implicitly. I might just replace TryRegister with Register and make no other changes.

**Comment by Shane32** [comment]
Yeah I don’t think we need to make any public api changes at this time.

**Comment by sungam3r** [comment]
> We can just cast to the underlying IServiceCollection and use TryAddEnumerable from there.

Yes but then it will work only sof MSDI.

> Server is already tied to MS DI for authentication

For auth only.

> Also since you have to manually call the builder method it’s not like it may have been called twice implicitly. I might just replace TryRegister with Register and make no other changes.

These are speculation. They can be true and may not.


TryRegister/Register  and Single/Enumerable registrations are completely independent concepts. In order not to combinatorial increase the number of methods, I propose to make an attribute like
```c#
public class EnumerableUsageAttribute : Attribute { }
```

Then check that attribute from GraphQL.NET. If found **either** on serviceType or implementationType then call appropriate method for enumerable. It potentially will work for any DI provider.

**Comment by sungam3r** [comment]
Public API will be complemented by an attribute, it is a backward compatible change.

**Comment by Shane32** [comment]
> These are speculation. They can be true and may not.

please explain.

**Comment by Shane32** [comment]
The problem is that attributes cannot be added to types outside the control of the 3rd party library.

**Comment by Shane32** [comment]
There’s no “enumerable” version of Register/Add. It will always add another registration. We don’t need to think about changes there. Only TryRegister differs from TryRegisterEnumerable. Even then, probably only the type overload, not the factory or instance overloads. 

Chances are that by adding TryRegisterEnumerable we are going from 6 registration methods to 7, not 6 to 12. I’d have to check but that’s what my gut is saying.

**Comment by sungam3r** [comment]
> The problem is that attributes cannot be added to types outside the control of the 3rd party library.

I know. At least now the attribute makes it possible to correct **our** error and provides ability to tune registrations further.

Most of the time I have been developing libraries, rather than end-user applications. My consumers are other developers. In my work, I constantly come across the fact that libraries are built into stacks. Under these conditions, it is important to be as attentive and not to make assumptions. One layer knows nothing about another layer under or on top of or "on the side".

**Comment by sungam3r** [comment]
![изображение](https://user-images.githubusercontent.com/21261007/148215057-290bcc22-3efc-42da-82b4-442452712fd8.png)
![изображение](https://user-images.githubusercontent.com/21261007/148215118-a1227e4a-8464-4fca-8c78-eb9f96c9d0b5.png)

So... 6 -> 9 😉

**Comment by sungam3r** [comment]
> There’s no “enumerable” version of Register/Add.

Agree. Let's tune TryRegister behavior.

**Comment by sungam3r** [comment]
Or add something like _replace_ argument for TryRegister.

**Comment by Shane32** [comment]
Exactly why attribute would be a bad design. It would require a lower layer to have knowledge of the GraphQL layer. 

If I needed to use this method it would break my own code already. Because I put services in a separate project from my GraphQL code with no reference to GraphQL. Now I would be forced to create a reference in order to add the attribute.

**Comment by Shane32** [comment]
> Or add something like _replace_ argument for TryRegister.

Sounds good. IsEnumerable perhaps

**Comment by sungam3r** [comment]
` IGraphQLBuilder TryRegister(Type serviceType, Type implementationType, ServiceLifetime serviceLifetime);`
->
` IGraphQLBuilder TryRegister(Type serviceType, Type implementationType, ServiceLifetime serviceLifetime, bool asEnumerable = false);`

**Comment by Shane32** [comment]
Right. Although in v4 it needs to be an overload so it is not a breaking change. In v5 it should be an optional argument

**Comment by Shane32** [comment]
Or We can change TryRegister to register on that one method for v4 and add the optional argument for v5

**Comment by sungam3r** [comment]
OK. I suggest to add `RegistrationCompareMode mode = RegistrationCompareMode.ServiceType` to make intentions clear:
```c#
public enum RegistrationCompareMode
{
   ServiceType,
   ServiceTypeAndImplementationType
}
```

`Enumerable` or some other type under the hood is implementation details.

**Comment by Shane32** [comment]
fine with me

**Comment by lanwin** [comment]
👍🏻  I have to say thank you for you both efforts! 

By the way. Did you noticed my inline comment :D -> 

     (it would be helpful if they had a description what todo to get rid of the obsolete)

**Comment by lanwin** [comment]
Do you have any plans when this fix will end up in a release?

**Comment by sungam3r** [comment]
I don't know yet, maybe for GraphQL.NET v5 release in April.

**Comment by sungam3r** [comment]
I came across this error when migrating my code as well. I actively use `TryAddEnumerable`. So the good news are that I have to fix this ASAP. Tomorrow, I think.

**Comment by sungam3r** [comment]
Fixed in #730.
